File: ts/src/manifest.ts
import {
  EDGE_LIVE_IN_TABLE,
  EDGE_LIVE_OUT_TABLE,
  EDGE_LOG_TABLE,
  GRAPH_FORMAT,
  VERTEX_LIVE_TABLE,
  VERTEX_LOG_TABLE,
} from "./schema.js";

export interface GraphManifestTableRef {
  'tableName': string;
  uri: string;
  version?: number;
  etag?: string | null;
}

export interface GraphManifestTables {
  'vertexLog': GraphManifestTableRef;
  'edgeLog': GraphManifestTableRef;
  'vertexLive': GraphManifestTableRef;
  'edgeLiveOut': GraphManifestTableRef;
  'edgeLiveIn': GraphManifestTableRef;
}

export interface GraphManifestSeqRange {
  min: number;
  max: number;
}

export interface GraphManifest {
  'manifestVersion': number;
  'graphFormat': typeof GRAPH_FORMAT;
  'contractVersion': number;
  'partitionId': number;
  version: number;
  'keyLayout': GraphManifestKeyLayout;
  tables: GraphManifestTables;
  seq: GraphManifestSeqRange;
  dirtyLabels?: string[];
  generatedAtMs?: number;
}

export interface GraphManifestKeyLayout {
  'manifestPrefix': string;
  'latestKey': string;
  'pointerKey': string;
  'versionedKey': string;
}

export interface GraphManifestLatestPointer {
  'graphFormat': typeof GRAPH_FORMAT;
  'contractVersion': number;
  'partitionId': number;
  version: number;
  'versionedKey': string;
  etag?: string | null;
  updatedAtMs?: number;
}

export interface ManifestStore {
  get(key: string): Promise<string | null>;
  put(key: string, value: string): Promise<void>;
}

export interface R2BucketLike {
  get(key: string): Promise<{ text(): Promise<string> } | null>;
  put(
    key: string,
    value: string,
    options?: {
      httpMetadata?: {
        contentType?: string;
      };
    },
  ): Promise<unknown>;
}

export function makeManifestTableRef(tableName: string, uri: string): GraphManifestTableRef {
  return { 'tableName': tableName, uri };
}

export function createGraphManifest(
  partitionId: number,
  version: number,
  seq: GraphManifestSeqRange,
  baseUri: string,
): GraphManifest {
  const base = baseUri.replace(/\/+$/, "");
  return {
    'manifestVersion': 1,
    'graphFormat': GRAPH_FORMAT,
    'contractVersion': 1,
    'partitionId': partitionId,
    version,
    'keyLayout': createManifestKeyLayout(partitionId, version),
    tables: {
      'vertexLog': makeManifestTableRef(VERTEX_LOG_TABLE, `${base}/${VERTEX_LOG_TABLE}`),
      'edgeLog': makeManifestTableRef(EDGE_LOG_TABLE, `${base}/${EDGE_LOG_TABLE}`),
      'vertexLive': makeManifestTableRef(VERTEX_LIVE_TABLE, `${base}/${VERTEX_LIVE_TABLE}`),
      'edgeLiveOut': makeManifestTableRef(EDGE_LIVE_OUT_TABLE, `${base}/${EDGE_LIVE_OUT_TABLE}`),
      'edgeLiveIn': makeManifestTableRef(EDGE_LIVE_IN_TABLE, `${base}/${EDGE_LIVE_IN_TABLE}`),
    },
    seq,
  };
}

export function createManifestKeyLayout(
  partitionId: number,
  version: number,
): GraphManifestKeyLayout {
  const manifestPrefix = `manifests/partitions/${partitionId}`;
  return {
    'manifestPrefix': manifestPrefix,
    'latestKey': `${manifestPrefix}/latest.json`,
    'pointerKey': `${manifestPrefix}/latest.json`,
    'versionedKey': `${manifestPrefix}/manifest-${String(version).padStart(20, "0")}.json`,
  };
}

export function stringifyManifest(manifest: GraphManifest): string {
  return JSON.stringify(manifest, null, 2);
}

export function parseManifest(json: string): GraphManifest {
  return JSON.parse(json) as GraphManifest;
}

export async function saveManifest(
  store: ManifestStore,
  key: string,
  manifest: GraphManifest,
): Promise<void> {
  await store.put(key, stringifyManifest(manifest));
}

export async function loadManifest(
  store: ManifestStore,
  key: string,
): Promise<GraphManifest | null> {
  const json = await store.get(key);
  return json == null ? null : parseManifest(json);
}

export async function loadLatestPointer(
  store: ManifestStore,
  key: string,
): Promise<GraphManifestLatestPointer | null> {
  const json = await store.get(key);
  return json == null ? null : parseLatestPointer(json);
}

export async function loadManifestViaLatest(
  store: ManifestStore,
  latestKey: string,
): Promise<GraphManifest | null> {
  const pointer = await loadLatestPointer(store, latestKey);
  return pointer == null ? null : loadManifest(store, pointer.versionedKey);
}

/**
 * Publish a manifest with atomic pointer update.
 *
 * Writes the immutable versioned manifest first, then updates the latest
 * pointer with retry + read-after-write verification to guard against
 * partial R2 writes.
 */
export async function publishManifest(
  store: ManifestStore,
  manifest: GraphManifest,
): Promise<void> {
  // 1. Write versioned manifest (immutable, idempotent)
  await saveManifest(store, manifest.keyLayout.versionedKey, manifest);

  // 2. Atomic pointer update with retry + read-after-write verify
  const pointerBody = stringifyLatestPointer(createLatestPointer(manifest));
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      await store.put(manifest.keyLayout.pointerKey, pointerBody);
      // Verify: read-after-write consistency check
      const verify = await store.get(manifest.keyLayout.pointerKey);
      if (verify) {
        try {
          const parsed = JSON.parse(verify);
          if (parsed.version === manifest.version) return;
        } catch { /* parse error, retry */ }
      }
    } catch (e) {
      if (attempt === 2) throw e;
    }
    // Exponential backoff
    await new Promise(r => setTimeout(r, 100 * Math.pow(2, attempt)));
  }
  console.error(`publishManifest: pointer verify failed after 3 attempts (v=${manifest.version})`);
}

export function createLatestPointer(
  manifest: GraphManifest,
): GraphManifestLatestPointer {
  return {
    'graphFormat': GRAPH_FORMAT,
    'contractVersion': manifest.contractVersion,
    'partitionId': manifest.partitionId,
    version: manifest.version,
    'versionedKey': manifest.keyLayout.versionedKey,
    etag: manifest.tables.vertexLive.etag ?? manifest.tables.edgeLiveOut.etag,
    'updatedAtMs': manifest.generatedAtMs,
  };
}

export function stringifyLatestPointer(pointer: GraphManifestLatestPointer): string {
  return JSON.stringify(pointer, null, 2);
}

export function parseLatestPointer(json: string): GraphManifestLatestPointer {
  return JSON.parse(json) as GraphManifestLatestPointer;
}

export function createFetchManifestStore(fetchImpl: typeof fetch = fetch): ManifestStore {
  return {
    async get(key: string): Promise<string | null> {
      const res = await fetchImpl(key, { method: "GET" });
      if (res.status === 404) return null;
      if (!res.ok) throw new Error(`manifest GET failed: ${res.status}`);
      return res.text();
    },
    async put(key: string, value: string): Promise<void> {
      const res = await fetchImpl(key, {
        method: "PUT",
        headers: { "content-type": "application/json" },
        body: value,
      });
      if (!res.ok) throw new Error(`manifest PUT failed: ${res.status}`);
    },
  };
}

export function createR2ManifestStore(
  bucket: R2BucketLike,
  prefix = "",
): ManifestStore {
  const normalizedPrefix = prefix.replace(/\/+$/, "");
  const objectKey = (key: string) => {
    const trimmed = key.replace(/^\/+/, "");
    return normalizedPrefix ? `${normalizedPrefix}/${trimmed}` : trimmed;
  };

  return {
    async get(key: string): Promise<string | null> {
      const obj = await bucket.get(objectKey(key));
      return obj ? obj.text() : null;
    },
    async put(key: string, value: string): Promise<void> {
      await bucket.put(objectKey(key), value, {
        httpMetadata: { contentType: "application/json" },
      });
    },
  };
}

File: yata-lance/src/manifest.rs
use serde::{Deserialize, Serialize};

use crate::schema::{
    EDGE_LIVE_IN_TABLE,
    EDGE_LIVE_OUT_TABLE,
    EDGE_LOG_TABLE,
    GRAPH_FORMAT,
    VERTEX_LIVE_TABLE,
    VERTEX_LOG_TABLE,
};
use crate::YataDataset;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct GraphManifestTableRef {
    pub table_name: String,
    pub uri: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub etag: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct GraphManifestTables {
    pub vertex_log: GraphManifestTableRef,
    pub edge_log: GraphManifestTableRef,
    pub vertex_live: GraphManifestTableRef,
    pub edge_live_out: GraphManifestTableRef,
    pub edge_live_in: GraphManifestTableRef,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct GraphManifestSeqRange {
    pub min: u64,
    pub max: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct GraphManifest {
    pub manifest_version: u32,
    pub graph_format: String,
    pub contract_version: u32,
    pub partition_id: u32,
    pub version: u64,
    pub tables: GraphManifestTables,
    pub seq: GraphManifestSeqRange,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub dirty_labels: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub generated_at_ms: Option<u64>,
}

pub struct OpenedGraphDatasets {
    pub vertex_log: YataDataset,
    pub edge_log: YataDataset,
    pub vertex_live: YataDataset,
    pub edge_live_out: YataDataset,
    pub edge_live_in: YataDataset,
}

pub trait ManifestStore {
    type Error;

    fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Self::Error>;
    fn put(&self, key: &str, value: &[u8]) -> Result<(), Self::Error>;
}

impl GraphManifest {
    pub fn new(partition_id: u32, version: u64, seq_min: u64, seq_max: u64, base_uri: &str) -> Self {
        let base_uri = base_uri.trim_end_matches('/');
        Self {
            manifest_version: 1,
            graph_format: GRAPH_FORMAT.to_string(),
            contract_version: 1,
            partition_id,
            version,
            tables: GraphManifestTables {
                vertex_log: GraphManifestTableRef::new(VERTEX_LOG_TABLE, format!("{base_uri}/{VERTEX_LOG_TABLE}")),
                edge_log: GraphManifestTableRef::new(EDGE_LOG_TABLE, format!("{base_uri}/{EDGE_LOG_TABLE}")),
                vertex_live: GraphManifestTableRef::new(VERTEX_LIVE_TABLE, format!("{base_uri}/{VERTEX_LIVE_TABLE}")),
                edge_live_out: GraphManifestTableRef::new(EDGE_LIVE_OUT_TABLE, format!("{base_uri}/{EDGE_LIVE_OUT_TABLE}")),
                edge_live_in: GraphManifestTableRef::new(EDGE_LIVE_IN_TABLE, format!("{base_uri}/{EDGE_LIVE_IN_TABLE}")),
            },
            seq: GraphManifestSeqRange { min: seq_min, max: seq_max },
            dirty_labels: Vec::new(),
            generated_at_ms: None,
        }
    }
}

impl GraphManifestTableRef {
    pub fn new(table_name: &str, uri: String) -> Self {
        Self {
            table_name: table_name.to_string(),
            uri,
            version: None,
            etag: None,
        }
    }
}

pub fn render_manifest_json(manifest: &GraphManifest) -> Result<Vec<u8>, serde_json::Error> {
    serde_json::to_vec_pretty(manifest)
}

pub fn parse_manifest_json(data: &[u8]) -> Result<GraphManifest, serde_json::Error> {
    serde_json::from_slice(data)
}

pub fn save_manifest<S: ManifestStore>(
    store: &S,
    key: &str,
    manifest: &GraphManifest,
) -> Result<(), String> {
    let bytes = render_manifest_json(manifest).map_err(|e| e.to_string())?;
    store.put(key, &bytes).map_err(|_| "manifest put failed".to_string())
}

pub fn load_manifest<S: ManifestStore>(store: &S, key: &str) -> Result<Option<GraphManifest>, String> {
    let bytes = match store.get(key).map_err(|_| "manifest get failed".to_string())? {
        Some(bytes) => bytes,
        None => return Ok(None),
    };
    parse_manifest_json(&bytes)
        .map(Some)
        .map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn manifest_roundtrip() {
        let mut manifest = GraphManifest::new(0, 7, 1, 99, "r2://bucket/yata/partitions/0");
        manifest.dirty_labels = vec!["Post".into(), "Profile".into()];
        manifest.generated_at_ms = Some(1_711_900_000_000);

        let bytes = render_manifest_json(&manifest).unwrap();
        let parsed = parse_manifest_json(&bytes).unwrap();
        assert_eq!(parsed, manifest);
    }
}
